tests/python/pants_test/pantsd/test_pailgun_server.py

```diff
@@ -78,6 +82,102 @@ def test_process_request_thread_error(self, mock_shutdown_request):
     self.assertIs(self.mock_handler_inst.handle_error.called, True)
     mock_shutdown_request.assert_called_once_with(self.server, mock_request)
 
+  def test_ensure_request_is_exclusive(self):
+    """Launch many requests, assert that every one is trying to enter the critical section, and assert that only one is doing so at a time."""
+    self.threads_to_start = 10
+
+    self.threads_running_cond = threading.Condition()
+    self.threads_running = 0
+    def handle_thread_tried_to_handle_request():
+      """Mark a thread as started, and block until every thread has been marked as starting."""
+      self.threads_running_cond.acquire()
+      self.threads_running += 1
+      if self.threads_running == self.threads_to_start:
+        self.threads_running_cond.notify_all()
+      else:
+        while not self.threads_running == self.threads_to_start:
+          self.threads_running_cond.wait()
+
+      self.assertEqual(self.threads_running, self.threads_to_start, "This thread is unblocked before all the threads had started.")
```

Comment:
You shouldn't assert on background threads, as the `join()` will succeed rather than raising the exception on the main thread. Instead consider something like setting a variable, or appending to a list (under a lock) from the child thread, and after you `join()`, check that that is `None`/empty

Reply:
Used a `Queue` (which is thread safe) to store exceptions, and assert that it's empty at the end.